### DISCUSSION TITLE
Will the 'sometimes' rule be supported in the future / partial updates?

### DISCUSSION THREAD

**Discussion by cristiancalara** [spatie/laravel-data#65] [Q&A]
I was wondering if the package plans on supporting the 'sometimes' rule in the future?

The main use case for the above would be to support partial API updates.

**Comment by cristiancalara** [comment]
I've tried a few approaches for the partial update feature:
- using `prepareForValidation` on the FormRequets to fill in the missing fields. This was not ideal as in some cases I needed to query a lot of data that wasn't actually getting used.
- in the DTO class, in the `fromRequest` method, I would pass in an existing model optionally, and default to the attributes on the model if the request did have them. Mainly the same issue as above.
-  I'm now testing an approach, where I have an "Undefined" class, which is added to all the properties on the DTO. And then, in the methods like `only`, `has`, etc., all the Undefined values are filtered out. As a simple example of how things look:

```php
class ProductData extends DataTransferObject
{
    public function __construct(
        public string|Undefined $name = new Undefined,
        public bool|Undefined $available = new Undefined,
    ) {
    }

    public static function fromRequest(ProductRequest $request): ProductData
    {
        return new self(...$request->only([
            'name',
            'available'
        ])->toArray());
    }
}
```

```php
class DataTransferObject
{
    public function collection(): Collection
    {
        $data = new Collection();
        $class = new ReflectionClass(static::class);
        $properties = $class->getProperties(ReflectionProperty::IS_PUBLIC);
        foreach ($properties as $reflectionProperty) {
            // Skip static properties
            if ($reflectionProperty->isStatic()) {
                continue;
            }

            $value = $reflectionProperty->getValue($this);
            // Skipping Undefined values.
            if ($value instanceof Undefined) {
                continue;
            }
            $data->put($reflectionProperty->getName(), $reflectionProperty->getValue($this));
        }

        return $data;
    }

    public function has($key): bool
    {
        return $this->collection()->has($key);
    }

    public function only($keys)
    {
        return $this->collection()->only($keys);
    }

    public function toArray(): array
    {
        return $this->collection()->toArray();
    }
}
```

Was not sure if I should add more details about the partial API update thing, as maybe that is not the scope of the package but I thought it might help the discussion.

**Comment by cristiancalara** [comment]
Played around a bit with this and created a proof of concept as fork https://github.com/cristiancalara/laravel-data/commit/bd327721991175d88abd0eb07626c56f10c1f713 (was not sure if I should create a PR without the feature being approved/wanted).

While I think adding Undefined to all types is making things a bit harder to read, it might be worth allowing this approach via a flag? This would allow having a 1 to 1 mapping between arrays and data objects - as currently, in an array, we have the concept of undefined, but it's missing in the data object.

**Comment by rubenvanassche** [comment]
Hi @cristiancalara,

I like the idea, though we need to make sure everything keeps working fine. Could you add the PR to this repo? Then I can check it out a bit more. We also need to make sure a `Sometimes` rule attribute will be added which works together with the `Undefined` class. Then we only need to add documentation and this would be a go for me!

**Comment by ragulka** [comment]
I bumped into something similar, I think. I allow partial (atomic) updates to models via my controller. The point is to follow the PATCH HTTP verb definition - not to replace an existing resource, but to patch one or more parts of it. This is very simply with simple arrays, as you can simply check if the key exists or not, but I'm trying to use the same with the Data object. The issue is that I need to convert the data object back to an array to update the record in DB. Something like this:

```php
class ProductController extends Controller {
    public function update(Product $product, ProductData $data)
    {
        $product->update($data->toArray());

        return redirect()->back()->with('success', __('Product successfully updated'));
    }
}
```

The problem there is that the request body may not contain all the fields, and so the undefined fields get overwritten with `NULL` values. The issue can be demonstrated by the following:

```php
class ProductData extends Data {
    public function __construct(
        public string $name,
        public ?string $description,
    ) {
    }
}

$data = ProductData::from(['name' => 'Foo']);

$data->toArray();  // will produce ['name' => 'Foo', 'description' => null];
```
I understand how the above makes sense, but it would be great if there was a way to exclude undefined properties altogether when converting back to array.

**Comment by ragulka** [comment]
Well, it turns out this feature has already been implemented, and is named `Optional`: https://spatie.be/docs/laravel-data/v2/as-a-data-transfer-object/optional-properties
